clang/lib/CIR/Dialect/IR/CIRTypes.cpp

```diff
@@ -932,6 +932,34 @@ llvm::ArrayRef<mlir::Type> FuncType::getReturnTypes() const {
 
 bool FuncType::isVoid() const { return mlir::isa<VoidType>(getReturnType()); }
 
+//===----------------------------------------------------------------------===//
+// MethodType Definitions
+//===----------------------------------------------------------------------===//
+
+llvm::TypeSize
+MethodType::getTypeSizeInBits(const mlir::DataLayout &dataLayout,
+                              mlir::DataLayoutEntryListRef params) const {
+  // TODO: consider size differences under different ABIs
```

Comment:
We already have a better ABI infra, you should implement what's necessary here, at least for one target.

Reply:
How should I query for ABI specific information in the data layout interface? Btw this problem also exists in ABI-related other types such as pointers and members.

Reply:
This could be more subtle. The Itanium ABI actually defines the layout of a member function pointer as something like `struct { fnptr_t, ptrdiff_t }` where `fnptr_t` is the type for normal function pointers. Thus on most 64-bit platforms a member function pointer will end up being a `struct { int64_t, int64_t }`, and its size is 128 bits, thus the default implementation `llvm::TypeSize::getFixed(128)`.

Three points worth mentioning here:
1. Nothing in the standard stops `ptrdiff_t` from having a different size or alignment requirement than a normal pointer. (but this is rare and I doubt if there are any implementations doing this)
2. On 32-bit platforms both of `fnptr_t` and `ptrdiff_t` become `int32_t` and the member function pointer size changes as well.
3. In MSVC ABI a member function pointer has the same layout as a single function pointer, which is totally different from Itanium ABI.

For now we're focusing on the Itanium ABI only, so maybe let's first tackle the second point. A better type size query implementation for the member function pointers would then be the size of a `struct { pointer, pointer }`.

Reply:
> For now we're focusing on the Itanium ABI only, so maybe let's first tackle the second point. A better type size query implementation for the member function pointers would then be the size of a `struct { pointer, pointer }`.

Sounds good to me, please change it to the size of a `struct { pointer, pointer }` and add a missing feature assert for MSVC ABI.

Reply:
Updated.